cdap-cli/src/main/java/co/cask/cdap/cli/util/AbstractAuthCommand.java

```diff
@@ -72,29 +76,38 @@ public void execute(Arguments arguments, PrintStream printStream) throws Excepti
   public abstract void perform(Arguments arguments, PrintStream printStream) throws Exception;
 
   protected ServiceId parseServiceId(Arguments arguments) {
-    String[] appVersionServiceId = arguments.get(ArgumentName.SERVICE.toString()).split("\\.");
-    if (appVersionServiceId.length < 2) {
-      throw new CommandInputError(this);
+    String[] idParts = parseProgramIdArgument(arguments, ArgumentName.SERVICE.getName());
+    if (idParts[VERSION_IDX] == null) {
+      idParts[VERSION_IDX] =  ApplicationId.DEFAULT_VERSION;
     }
+    return cliConfig.getCurrentNamespace().app(idParts[APP_IDX], idParts[VERSION_IDX]).service(idParts[PROGRAM_IDX]);
+  }
 
-    String appId = appVersionServiceId[0];
-    String serviceId = appVersionServiceId[appVersionServiceId.length - 1];
-    String appVersion = appVersionServiceId.length == 2 ? ApplicationId.DEFAULT_VERSION :
-      dotJoiner.join(Arrays.copyOfRange(appVersionServiceId, 1, appVersionServiceId.length - 1));
-    return cliConfig.getCurrentNamespace().app(appId, appVersion).service(serviceId);
+  protected ServiceId parseNonversionServiceId(Arguments arguments) {
+    String[] idParts = parseProgramIdArgument(arguments, ArgumentName.NON_VERSION_SERVICE.getName());
+    return cliConfig.getCurrentNamespace().app(idParts[APP_IDX]).service(idParts[PROGRAM_IDX]);
   }
 
   protected ProgramId parseProgramId(Arguments arguments, ElementType elementType) {
-    String[] programIdParts = arguments.get(elementType.getArgumentName().toString()).split("\\.");
-    if (programIdParts.length < 2) {
+    String[] idParts = parseProgramIdArgument(arguments, elementType.getArgumentName().getName());
+    if (idParts[VERSION_IDX] == null) {
+      idParts[VERSION_IDX] =  ApplicationId.DEFAULT_VERSION;
+    }
+    return cliConfig.getCurrentNamespace().app(idParts[APP_IDX], idParts[VERSION_IDX])
+      .program(elementType.getProgramType(), idParts[PROGRAM_IDX]);
+  }
+
+  protected String[] parseProgramIdArgument(Arguments arguments, String argumentName) {
```

Comment:
Would it be cleaner to return an object from here? Maybe call it ProgramIdParts, and store application, version and program name in it? It would make the code much more readable than using indexes in an array.